Backpack/Dialog/Classes/BPKDialogController.m

```diff
@@ -0,0 +1,224 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "BPKDialogController.h"
+
+#import "BPKDialogButtonAction.h"
+#import "BPKDialogScrimAction.h"
+#import "BPKDialogControllerAnimator.h"
+
+#import <Backpack/Color.h>
+#import <Backpack/Spacing.h>
+#import <Backpack/Font.h>
+#import <Backpack/Color.h>
+#import <Backpack/Button.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BPKDialogController()
+
+@property(nonatomic, strong) UIColor *headColor;
+@property(nonatomic, strong) UIImage *iconImage;
+@property(nonatomic, copy) NSString *titleText;
+@property(nonatomic, copy) NSString *messageText;
+
+@property(nonatomic, strong) UIView *scrimView;
+@property(nonatomic, strong) BPKDialogView *dialogView;
+
+@property(nonatomic, strong) BPKDialogScrimAction *scrimAction;
+
+@property(nonatomic, strong) NSLayoutConstraint *bottomAnchorConstraint;
+@end
+
+@implementation BPKDialogController
+
+- (instancetype)initWithTitle:(nullable NSString *)title
+                      message:(nullable NSString *)message
+                        style:(BPKDialogControllerStyle)style
+                  shadowStyle:(BPKDialogControllerShadowStyle)shadowStyle
+                    headColor:(UIColor *)headColor
+                    iconImage:(UIImage *)iconImage {
+    self = [super init];
+
+    if (self) {
+        self.titleText = title;
+        self.messageText = message;
+        self.style = style;
+        self.shadowStyle = shadowStyle;
+        self.headColor = headColor;
+        self.iconImage = iconImage;
+        self.transitioningDelegate = self;
+
+        [self setupViews];
+        [self addViews];
+        [self setupConstraints];
+    }
+    
+    return self;
+}
+
++ (instancetype)dialogControllerWithTitle:(NSString *)title
+                                  message:(NSString *)message
+                                    style:(BPKDialogControllerStyle)style
+                              shadowStyle:(BPKDialogControllerShadowStyle)shadowStyle
+                                headColor:(UIColor *)headColor
+                                iconImage:(UIImage *)iconImage {
+    return [[self alloc] initWithTitle:title message:message style:style shadowStyle:shadowStyle headColor:headColor iconImage:iconImage];
+}
+
+- (void)setupViews {
+    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(scrimTapped:)];
+    
+    self.scrimView = [UIView new];
```

Comment:
sometimes you use alloc + initWithFrame:CGRectZero, sometimes new

Reply:
I think I use `initWithFrame:CGRectZero` pretty consistently, but this code in particular was written by both me and @zsbee which is why it's a bit inconsistent.